moodle/podcontroller.go

```diff
@@ -0,0 +1,242 @@
+// workqueue.
+func (c *PodController) runWorker() {
+	for c.processPodQueueItem() {
+	}
+}
+func (c *PodController) processPodQueueItem() bool {
+	podObj, shutdown := c.podqueue.Get()
+	if shutdown {
+		return false
+	}
+	err := func(podObj interface{}) error {
+		defer c.podqueue.Done(podObj)
+		var key string
+		var ok bool
+		if key, ok = podObj.(string); !ok {
+			c.podqueue.Forget(podObj)
+			runtime.HandleError(fmt.Errorf("PodController.go: expected string in podqueue but got %#v", podObj))
+			return nil
+		}
+		if err := c.handlePod(key); err != nil {
+			return fmt.Errorf("PodController.go: error in handlePod key: %s, err: %s", key, err.Error())
+		}
+		c.podqueue.Forget(podObj)
+		return nil
+	}(podObj)
+
+	if err != nil {
+		runtime.HandleError(err)
+		return true
+	}
+	return true
+}
+
+// enqueuePod takes a Pod resource and converts it into a namespace/name
+// string which is then put onto the work queue. This method should *not* be
+// passed resources of any type other than Foo.
+func (c *PodController) enqueuePod(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	fmt.Printf("PodController.go     : enqueuePod, adding a pod key %s\n", key)
+	c.podqueue.AddRateLimited(key)
+}
+func (c *PodController) handlePod(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("PodController.go: invalid resource key, can't split : %s", key))
+		return fmt.Errorf("PodController.go: invalid resource key, can't split : %s", key)
+	}
+	pod, err := c.kubeclientset.CoreV1().Pods(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		// fmt.Println("PodController.go     : Couldn't find the replicaset, skipping")
+		return fmt.Errorf("Could not find the pod for pod name : %s, namespace %s", name, namespace)
+	}
+	ownerReferences := pod.GetOwnerReferences()
+	if len(ownerReferences) == 0 {
+		// fmt.Println("PodController.go     : Pod does not have any owner references.")
+		return nil
+	}
+	ownerReferenceName := ownerReferences[0].Name
+
+	replicaSet, err := c.kubeclientset.AppsV1().ReplicaSets(namespace).Get(ownerReferenceName, metav1.GetOptions{})
+	if err != nil {
+		// fmt.Println("PodController.go     : Couldn't find the replicaset, skipping")
+		return nil
+	}
+	ownerReferences = replicaSet.GetOwnerReferences()
+	if len(ownerReferences) == 0 {
+		// fmt.Println("PodController.go     : replicaset does not have a ny owner references.")
+		return nil
+	}
+	ownerReferenceName = ownerReferences[0].Name
+
+	deployment, err := c.kubeclientset.AppsV1().Deployments(namespace).Get(ownerReferenceName, metav1.GetOptions{})
+	if err != nil {
+		// fmt.Println("PodController.go     : Couldn't find the deployment, skipping")
+		return nil
```

Comment:
Same comment as above. Return formatted error and not nil.

Reply:
same as above

Reply:
Make sense.